finance_dl/schwab.py

```diff
@@ -0,0 +1,303 @@
+                logger.info(f"Downloaded {dest_path}")
+                return True
+        return False
+
+    def get_num_transaction_types(self) -> int:
+        filter_link, = self.get_elements_wait("a.transaction-search-link")
+        filter_link.click()
+
+        checkbox_div, = self.get_elements_wait("div.transaction-filter-checkbox")
+
+        checkboxes = checkbox_div.find_elements(By.CSS_SELECTOR, "span")
+
+        modal_close, = self.get_elements_wait("button#modalClose")
+        modal_close.click()
+
+        return len(checkboxes)
+
+    def select_next_unseen_account(self, seen: Set[Account]) -> Optional[Account]:
+        for link in self.get_account_links():
+            account = self.get_account_from_container(link)
+
+            if account is None:
+                continue
+
+            if account in seen:
+                continue
+
+            link.click()
+
+            def ready(driver):
+                sel = self.get_elements_wait("button.account-selector-button")
+                if sel:
+                    acct = self.get_account_from_container(sel[0])
+                    if acct is not None:
+                        return acct.number == account.number
+                return False
+
+            WebDriverWait(self.driver, 30).until(
+                EC.invisibility_of_element_located(
+                    (By.CSS_SELECTOR, "sdps-account-selector__header")))
+            # Make sure Schwab registers the selection of the new account.
+            time.sleep(1)
+
+            return account
+        return None
+
+    def get_account_links(self) -> List[Any]:
+        selector_button, = self.get_elements_wait("button.account-selector-button")
+        selector_button.click()
+
+        return self.get_elements_wait(
+            "li.sdps-account-selector__list-item a"
+        )
+
+    def get_account_from_container(self, container: Any) -> Optional[Account]:
+        spans = container.find_elements(By.CSS_SELECTOR, "span")
+        if len(spans) < 2:
+            return None
+        label_span, number_span = spans[:2]
+        number = number_span.text
+        if not number:
+            return None
+        return Account(label=label_span.text, number=number)
+
+    def login_if_needed(self) -> None:
+        login_frames = []
+        logout_buttons = []
+
+        def predicate(driver):
+            login_frames.extend(
+                driver.find_elements(By.CSS_SELECTOR, "iframe#lmsSecondaryLogin")
+            )
+            logout_buttons.extend(
+                driver.find_elements(By.CSS_SELECTOR, "button.logout")
+            )
+            return login_frames or logout_buttons
+
+        WebDriverWait(self.driver, 30).until(
+            predicate, message="Did not find either app or login page."
+        )
+
+        if logout_buttons:
+            logger.info("Already logged in.")
+            return
+
+        (login_frame,) = login_frames
+
+        logger.info("Logging in.")
```

Comment:
You might find you can simplify this with: `self.find_username_and_password_in_any_frame`

Reply:
Looked at this also, but it would require a pretty significant change to the way the login flow is currently implemented, which accounts for the possibility that we are already logged in in the browser session. Doesn't seem to bring enough obvious benefit to justify reworking that.